src/test/groovy/graphql/servlet/DataLoaderDispatchingSpec.groovy

```diff
@@ -0,0 +1,149 @@
+package graphql.servlet
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import graphql.execution.reactive.SingleSubscriberPublisher
+import graphql.schema.DataFetcher
+import graphql.schema.DataFetchingEnvironment
+import graphql.servlet.context.DefaultGraphQLServletContext
+import graphql.servlet.context.GraphQLContext
+import graphql.servlet.context.GraphQLContextBuilder
+import graphql.servlet.context.ContextSetting
+import org.dataloader.BatchLoader
+import org.dataloader.DataLoader
+import org.dataloader.DataLoaderRegistry
+import org.springframework.mock.web.MockHttpServletRequest
+import org.springframework.mock.web.MockHttpServletResponse
+import spock.lang.Shared
+import spock.lang.Specification
+
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+import javax.servlet.http.Part
+import javax.websocket.Session
+import javax.websocket.server.HandshakeRequest
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+
+class DataLoaderDispatchingSpec extends Specification {
+
+    public static final int STATUS_OK = 200
+    public static final String CONTENT_TYPE_JSON_UTF8 = 'application/json;charset=UTF-8'
+
+    @Shared
+    ObjectMapper mapper = new ObjectMapper()
+
+    AbstractGraphQLHttpServlet servlet
+    MockHttpServletRequest request
+    MockHttpServletResponse response
+    AtomicInteger fetchCounter = new AtomicInteger()
+    AtomicInteger loadCounter = new AtomicInteger()
+
+    BatchLoader<String, String> batchLoaderA = new BatchLoader<String, String>() {
+        @Override
+        CompletionStage<List<String>> load(List<String> keys) {
+            fetchCounter.incrementAndGet()
+            CompletableFuture.completedFuture(keys)
+        }
+    }
+
+    def registry() {
+        DataLoaderRegistry registry = new DataLoaderRegistry()
+        registry.register("A", DataLoader.newDataLoader(batchLoaderA))
+        registry
+    }
+
+    def setup() {
+        request = new MockHttpServletRequest()
+        request.setAsyncSupported(true)
+        request.asyncSupported = true
+        response = new MockHttpServletResponse()
+    }
+
+    def queryDataFetcher() {
+        return new DataFetcher() {
+            @Override
+            Object get(DataFetchingEnvironment environment) {
+                String id = environment.arguments.arg
+                loadCounter.incrementAndGet()
+                environment.getDataLoader("A").load(id)
+            }
+        }
+    }
+
+    def contextBuilder () {
+        return new GraphQLContextBuilder() {
+            @Override
+            GraphQLContext build(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Map<String, List<Part>> fileParts) {
+                new DefaultGraphQLServletContext(registry(), null)
```

Comment:
These tests are currently the only reason why `DefaultGraphQLServletContext` needs to be `public`. Preferrably we can make that implementation class package private. Also it seems currently the context cannot be built using the builder to pass in the `registry` and `subject`. I'm guessing that part is still work in progress?

Reply:
I initially planned just the one static method that required the registry/subject but added the other and hadn't yet added the with method. Will update. Previously the context objects were public, I could see being able to extend it and not having to rewrite the base interface being useful, do we not want to allow that?

Reply:
Ah right. but then we need to make the constructor `protected` instead of `private`. I tend to try to keep things package private usually so it's not part of the public api and I'm free to mess around with it without breaking anything for consumers. But in this case they're in fact very simple POJO's so might as well make them public.